src/colmap/estimators/covariance.cc

```diff
@@ -0,0 +1,582 @@
+// Copyright (c) 2023, ETH Zurich and UNC Chapel Hill.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+//     * Neither the name of ETH Zurich and UNC Chapel Hill nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "colmap/estimators/covariance.h"
+
+#include "colmap/estimators/cost_functions.h"
+
+#include <ceres/crs_matrix.h>
+
+namespace colmap {
+
+BundleAdjustmentCovarianceEstimatorBase::
+    BundleAdjustmentCovarianceEstimatorBase(ceres::Problem* problem,
+                                            Reconstruction* reconstruction)
+    : problem_(THROW_CHECK_NOTNULL(problem)),
+      reconstruction_(THROW_CHECK_NOTNULL(reconstruction)) {
+  SetUp(reconstruction_);
+}
+
+BundleAdjustmentCovarianceEstimatorBase::
+    BundleAdjustmentCovarianceEstimatorBase(
+        ceres::Problem* problem,
+        const std::vector<const double*>& pose_blocks,
+        const std::vector<const double*>& point_blocks)
+    : problem_(THROW_CHECK_NOTNULL(problem)) {
+  SetUp(pose_blocks, point_blocks);
+}
+
+void BundleAdjustmentCovarianceEstimatorBase::SetUp(
```

Comment:
nit: is there a reason to not directly do this in the constructor instead of a separate method?

Reply:
No reasons anymore. I have updated this.